In [1]:
import pandas as pd
import numpy as np
import load_data
import clean_data
import transform_data

In [2]:
datos = load_data.leedatos("interrupcion-legal-del-embarazo.csv")

In [3]:
datos.head()

AÑO     MES  CVE_HOSPITAL  FINGRESO        AUTOREF EDOCIVIL_DESCRIPCION  \
0  2016   Enero            10  12/01/16  AUTORREFERIDA              soltera   
1  2016   Abril             1  25/04/16             no          unión libre   
2  2018   Julio             4  11/07/18            NaN              soltera   
3  2016   Enero            10  07/01/16  AUTORREFERIDA               casada   
4  2018  Agosto             7       NaN             no          unión libre   

   EDAD DESC_DERECHOHAB              NIVEL_EDU  \
0    34         ninguna      primaria completa   
1    27         ninguna  licenciatura completa   
2    29         ninguna    secundaria completa   
3    33         ninguna  preparatoria completa   
4    23             N/E  preparatoria completa   

                        OCUPACION  ... C_NUM                  MOTILES  \
0                     Desempleada  ...   2.0  interrupción voluntaria   
1                          Mesera  ...   NaN  interrupción voluntaria   
2  Trabajadora del Sector Público  ...   2.0  interrupción voluntaria   
3                     Ama de Casa  ...   2.0  interrupción voluntaria   
4                             NaN  ...   2.0  interrupción voluntaria   

  H_FINGRESO       DESC_SERVICIO  P_SEMGEST  P_DIASGESTA P_CONSENT  \
0         NE  Salud Reproductiva        0.0          NaN        Si   
1   25/04/16                 ILE       11.0          2.0        Si   
2         NE                 ILE        7.0          6.0       NaN   
3         NE  Salud Reproductiva        4.0          NaN        Si   
4   24/08/18  Gineco Obstetrícia        7.0          3.0       NaN   

                                     PROCILE  S_COMPLICA          PANTICONCEP  
0                                Misoprostol          Si                  NaN  
1                                 Aspiración          No               condón  
2  Mifepristona 200 mg + Misoprostol 800 mcg         NaN  Anticonceptivo Oral  
3                                Misoprostol          Si                  NaN  
4                                Misoprostol         NaN  implante subdermico  

[5 rows x 36 columns]

In [4]:
datos_format = clean_data.clean_column_headers(datos)
datos_format.head()

año     mes  cve_hospital  fingreso        autoref edocivil_descripcion  \
0  2016   Enero            10  12/01/16  AUTORREFERIDA              soltera   
1  2016   Abril             1  25/04/16             no          unión libre   
2  2018   Julio             4  11/07/18            NaN              soltera   
3  2016   Enero            10  07/01/16  AUTORREFERIDA               casada   
4  2018  Agosto             7       NaN             no          unión libre   

   edad desc_derechohab              nivel_edu  \
0    34         ninguna      primaria completa   
1    27         ninguna  licenciatura completa   
2    29         ninguna    secundaria completa   
3    33         ninguna  preparatoria completa   
4    23             N/E  preparatoria completa   

                        ocupacion  ... c_num                  motiles  \
0                     Desempleada  ...   2.0  interrupción voluntaria   
1                          Mesera  ...   NaN  interrupción voluntaria   
2  Trabajadora del Sector Público  ...   2.0  interrupción voluntaria   
3                     Ama de Casa  ...   2.0  interrupción voluntaria   
4                             NaN  ...   2.0  interrupción voluntaria   

  h_fingreso       desc_servicio  p_semgest  p_diasgesta p_consent  \
0         NE  Salud Reproductiva        0.0          NaN        Si   
1   25/04/16                 ILE       11.0          2.0        Si   
2         NE                 ILE        7.0          6.0       NaN   
3         NE  Salud Reproductiva        4.0          NaN        Si   
4   24/08/18  Gineco Obstetrícia        7.0          3.0       NaN   

                                     procile  s_complica          panticoncep  
0                                Misoprostol          Si                  NaN  
1                                 Aspiración          No               condón  
2  Mifepristona 200 mg + Misoprostol 800 mcg         NaN  Anticonceptivo Oral  
3                                Misoprostol          Si                  NaN  
4                                Misoprostol         NaN  implante subdermico  

[5 rows x 36 columns]

In [5]:
datos_lower = transform_data.data_to_lowercase(datos_format)
datos_lower.head()

año     mes  cve_hospital  fingreso        autoref edocivil_descripcion  \
0  2016   enero            10  12/01/16  autorreferida              soltera   
1  2016   abril             1  25/04/16             no          unión libre   
2  2018   julio             4  11/07/18            NaN              soltera   
3  2016   enero            10  07/01/16  autorreferida               casada   
4  2018  agosto             7       NaN             no          unión libre   

   edad desc_derechohab              nivel_edu  \
0    34         ninguna      primaria completa   
1    27         ninguna  licenciatura completa   
2    29         ninguna    secundaria completa   
3    33         ninguna  preparatoria completa   
4    23             n/e  preparatoria completa   

                        ocupacion  ... c_num                  motiles  \
0                     desempleada  ...   2.0  interrupción voluntaria   
1                          mesera  ...   NaN  interrupción voluntaria   
2  trabajadora del sector público  ...   2.0  interrupción voluntaria   
3                     ama de casa  ...   2.0  interrupción voluntaria   
4                             NaN  ...   2.0  interrupción voluntaria   

  h_fingreso       desc_servicio  p_semgest  p_diasgesta p_consent  \
0         ne  salud reproductiva        0.0          NaN        si   
1   25/04/16                 ile       11.0          2.0        si   
2         ne                 ile        7.0          6.0       NaN   
3         ne  salud reproductiva        4.0          NaN        si   
4   24/08/18  gineco obstetrícia        7.0          3.0       NaN   

                                     procile  s_complica          panticoncep  
0                                misoprostol          si                  NaN  
1                                 aspiración          no               condón  
2  mifepristona 200 mg + misoprostol 800 mcg         NaN  anticonceptivo oral  
3                                misoprostol          si                  NaN  
4                                misoprostol         NaN  implante subdermico  

[5 rows x 36 columns]

In [6]:
##quitar acentos en los datos
#df_sinacentos = datos_lower["s_complica"].replace("á","a").replace("é","e").replace("í","i").\
#    replace("ó","o").replace("ú","u")

In [22]:
#df_char = datos_lower.select_dtypes(exclude=np.number)
#encabezados_char = df_char.columns.values
df_sin_acentos = datos_lower.replace({"á": "a", "ó": "o","ú": "u"},regex=True)

In [9]:
encabezados_char

array(['mes', 'fingreso', 'autoref', 'edocivil_descripcion',
       'desc_derechohab', 'nivel_edu', 'ocupacion', 'religion',
       'parentesco', 'entidad', 'alc_o_municipio', 'fmenstrua',
       'consejeria', 'anticonceptivo', 'motiles', 'h_fingreso',
       'desc_servicio', 'p_consent', 'procile', 's_complica',
       'panticoncep'], dtype=object)

In [23]:
#df_sinacentos = transform_data.data_sin_acentos(datos_lower)
df_sinacentos.head()

año     mes  cve_hospital  fingreso        autoref edocivil_descripcion  \
0  2016   enero            10  12/01/16  autorreferida              soltera   
1  2016   abril             1  25/04/16             no          unión libre   
2  2018   julio             4  11/07/18            NaN              soltera   
3  2016   enero            10  07/01/16  autorreferida               casada   
4  2018  agosto             7       NaN             no          unión libre   

   edad desc_derechohab              nivel_edu  \
0    34         ninguna      primaria completa   
1    27         ninguna  licenciatura completa   
2    29         ninguna    secundaria completa   
3    33         ninguna  preparatoria completa   
4    23             n/e  preparatoria completa   

                        ocupacion  ... c_num                  motiles  \
0                     desempleada  ...   2.0  interrupción voluntaria   
1                          mesera  ...   NaN  interrupción voluntaria   
2  trabajadora del sector público  ...   2.0  interrupción voluntaria   
3                     ama de casa  ...   2.0  interrupción voluntaria   
4                             NaN  ...   2.0  interrupción voluntaria   

  h_fingreso       desc_servicio  p_semgest  p_diasgesta p_consent  \
0         ne  salud reproductiva        0.0          NaN        si   
1   25/04/16                 ile       11.0          2.0        si   
2         ne                 ile        7.0          6.0       NaN   
3         ne  salud reproductiva        4.0          NaN        si   
4   24/08/18  gineco obstetrícia        7.0          3.0       NaN   

                                     procile  s_complica          panticoncep  
0                                misoprostol          si                  NaN  
1                                 aspiración          no               condón  
2  mifepristona 200 mg + misoprostol 800 mcg         NaN  anticonceptivo oral  
3                                misoprostol          si                  NaN  
4                                misoprostol         NaN  implante subdermico  

[5 rows x 36 columns]

In [8]:
df_sinacentos["s_complica"].unique()

array(['si', 'no', nan, 'sí'], dtype=object)

### Ajustar para quitar ñ y acentos

In [8]:
#datos_new = transform_data.split_column(datos_lower,"geo_point","latitud","longitud")
#datos_new.head()


In [9]:
datos_eda = datos_lower
#datos_eda = datos_new.drop(columns = ["geo_point","geo_shape","nomgeo","gid"])
datos_eda.head()

año     mes  cve_hospital  fingreso        autoref edocivil_descripcion  \
0  2016   enero            10  12/01/16  autorreferida              soltera   
1  2016   abril             1  25/04/16             no          unión libre   
2  2018   julio             4  11/07/18            NaN              soltera   
3  2016   enero            10  07/01/16  autorreferida               casada   
4  2018  agosto             7       NaN             no          unión libre   

   edad desc_derechohab              nivel_edu  \
0    34         ninguna      primaria completa   
1    27         ninguna  licenciatura completa   
2    29         ninguna    secundaria completa   
3    33         ninguna  preparatoria completa   
4    23             n/e  preparatoria completa   

                        ocupacion  ... c_num                  motiles  \
0                     desempleada  ...   2.0  interrupción voluntaria   
1                          mesera  ...   NaN  interrupción voluntaria   
2  trabajadora del sector público  ...   2.0  interrupción voluntaria   
3                     ama de casa  ...   2.0  interrupción voluntaria   
4                             NaN  ...   2.0  interrupción voluntaria   

  h_fingreso       desc_servicio  p_semgest  p_diasgesta p_consent  \
0         ne  salud reproductiva        0.0          NaN        si   
1   25/04/16                 ile       11.0          2.0        si   
2         ne                 ile        7.0          6.0       NaN   
3         ne  salud reproductiva        4.0          NaN        si   
4   24/08/18  gineco obstetrícia        7.0          3.0       NaN   

                                     procile  s_complica          panticoncep  
0                                misoprostol          si                  NaN  
1                                 aspiración          no               condón  
2  mifepristona 200 mg + misoprostol 800 mcg         NaN  anticonceptivo oral  
3                                misoprostol          si                  NaN  
4                                misoprostol         NaN  implante subdermico  

[5 rows x 36 columns]

### 

In [10]:
datos_eda.shape

(62003, 36)

In [11]:
datos_eda.describe().transpose()

count         mean        std     min     25%     50%     75%  \
año           62003.0  2017.279519   1.038147  2016.0  2016.0  2017.0  2018.0   
cve_hospital  62003.0     4.957179   3.445688     1.0     2.0     4.0     7.0   
edad          62003.0    25.695450   6.280193    11.0    21.0    24.0    30.0   
menarca       59114.0    12.640018   1.753795     1.0    12.0    12.0    14.0   
fsexual       57750.0    16.978632   2.460949     0.0    15.0    17.0    18.0   
sememb        58403.0     7.439771   2.368694     0.0     6.0     7.0     9.0   
nhijos        58794.0     1.033966   1.141758     0.0     0.0     1.0     2.0   
gesta         59341.0     2.344922   1.455347     0.0     1.0     2.0     3.0   
naborto       57888.0     0.176306   0.471237     0.0     0.0     0.0     0.0   
npartos       58240.0     0.711624   1.066238     0.0     0.0     0.0     1.0   
ncesarea      57904.0     0.314935   0.600371     0.0     0.0     0.0     0.0   
nile          49598.0     0.173616   0.462694     0.0     0.0     0.0     0.0   
c_num         41725.0     1.683954   0.684814     0.0     1.0     2.0     2.0   
p_semgest     61974.0     6.842515   2.178380     0.0     5.0     7.0     8.0   
p_diasgesta   48602.0     5.179334  10.427514     0.0     2.0     3.0     5.0   

                 max  
año           2019.0  
cve_hospital    14.0  
edad            53.0  
menarca         30.0  
fsexual         54.0  
sememb          50.0  
nhijos          10.0  
gesta           47.0  
naborto         10.0  
npartos         12.0  
ncesarea         7.0  
nile             9.0  
c_num           26.0  
p_semgest       12.0  
p_diasgesta     94.0

Ya podemos observar que solo las variables año, cve_hospital y edad no tienen observaciones faltantes, ya que en count, su valor es igual al número total de observaciones (62003).



In [14]:
datos_eda.isna().sum()

año                         0
mes                         0
cve_hospital                0
fingreso                 1583
autoref                 45129
edocivil_descripcion      116
edad                        0
desc_derechohab          7041
nivel_edu                1284
ocupacion                7721
religion                 1820
parentesco              41840
entidad                    20
alc_o_municipio          4093
menarca                  2889
fsexual                  4253
fmenstrua                6704
sememb                   3600
nhijos                   3209
gesta                    2662
naborto                  4115
npartos                  3763
ncesarea                 4099
nile                    12405
consejeria               1956
anticonceptivo           6706
c_num                   20278
motiles                  1088
h_fingreso               7537
desc_servicio            4215
p_semgest                  29
p_diasgesta             13401
p_consent               44147
procile   

Veamos ahora cuántos valores distintos contiene cada columna

In [12]:
datos_eda.nunique()

año                        4
mes                       12
cve_hospital              14
fingreso                 936
autoref                   33
edocivil_descripcion       7
edad                      43
desc_derechohab           34
nivel_edu                 17
ocupacion                 53
religion                  13
parentesco                40
entidad                   32
alc_o_municipio          499
menarca                   27
fsexual                   44
fmenstrua               1508
sememb                   140
nhijos                    11
gesta                     26
naborto                   10
npartos                   11
ncesarea                   7
nile                       8
consejeria                 3
anticonceptivo            20
c_num                     10
motiles                   13
h_fingreso               582
desc_servicio              8
p_semgest                 13
p_diasgesta               82
p_consent                  2
procile                   20
s_complica    

Vemos que algunas variables tienen un número bastante limitado de valores posibles, por lo que bien pueden ser categóricas. Es el caso de *año, mes,cve_hospital,edocivil_descripcion, nivel_edu, religion, entidad,s_complica* e *indice_des*.

Veamos de qué tipo son actualmente:

In [14]:
datos_eda["religion"].unique()

array(['ninguna', 'católica', 'cristiana', nan, 'creyente', 'mormon',
       'otra', 'testigos de jeohva', 'evangélica',
       'adventista del séptimo día', 'agnóstica', 'budista', 'judía',
       'protestante'], dtype=object)

In [13]:
datos_eda.dtypes

año                       int64
mes                      object
cve_hospital              int64
fingreso                 object
autoref                  object
edocivil_descripcion     object
edad                      int64
desc_derechohab          object
nivel_edu                object
ocupacion                object
religion                 object
parentesco               object
entidad                  object
alc_o_municipio          object
menarca                 float64
fsexual                 float64
fmenstrua                object
sememb                  float64
nhijos                  float64
gesta                   float64
naborto                 float64
npartos                 float64
ncesarea                float64
nile                    float64
consejeria               object
anticonceptivo           object
c_num                   float64
motiles                  object
h_fingreso               object
desc_servicio            object
p_semgest               float64
p_diasge

Vemos que todas ellas, salvo *anio* y *bimestre*, ya son de tipo object. Convirtamos éstas:

In [11]:
#categoricas =  ["anio","bimestre"]
data_eda = transform_data.change_type(datos_eda,categoricas)
data_eda.dtypes

consumo_total_mixto     float64
anio                     object
consumo_prom_dom        float64
consumo_total_dom       float64
alcaldia                 object
colonia                  object
consumo_prom_mixto      float64
consumo_total           float64
consumo_prom            float64
consumo_prom_no_dom     float64
bimestre                 object
consumo_total_no_dom    float64
indice_des               object
latitud                 float64
longitud                float64
dtype: object

### Variables categóricas

Nos centraremos ahora en las variables categóricas del data frame, mismas que seleccionamos con:

In [16]:
data_eda_cat = data_eda.select_dtypes(exclude=np.number)

NameError: name 'data_eda' is not defined

Queremos analizar lo siguiente:
* Número de categorías
* Valor de las categorías
* Moda
* Histograma
* Valores faltantes
* Número de observaciones con valores faltantes
* Proporción de observaciones por categoría
* ¿Hay faltas de ortografía?

### Variables numéricas

Nos centraremos ahora en las variables numéricas del data frame, mismas que seleccionamos con:

In [13]:
data_eda_num = data_eda.select_dtypes(include=np.number)

Queremos analizar lo siguiente:
* Tipo de dato: float, integer
* Número de observaciones
* Mean
* Desviación estándar
* Cuartiles: 25%, 50%, 75%
* Valor máximo
* Valor mínimo
* Número de observaciones únicos
* Top 5 observaciones repetidas
* Número de observaciones con valores faltantes
* ¿Hay redondeos?

### Feature selection & engineering

In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer

Revisamos los datos faltantes:

In [15]:
datos_eda.isna().sum()

consumo_total_mixto     8327
anio                       0
consumo_prom_dom        4820
consumo_total_dom       4820
alcaldia                   0
colonia                    0
consumo_prom_mixto      8327
consumo_total              0
consumo_prom               0
consumo_prom_no_dom        0
bimestre                   0
consumo_total_no_dom       0
indice_des                 0
latitud                    0
longitud                   0
dtype: int64

Haremos la imputaciones a los datos faltantes, usando la mediana:

In [24]:
datos_eda.dtypes

consumo_total_mixto     float64
anio                      int64
consumo_prom_dom        float64
consumo_total_dom       float64
alcaldia                 object
colonia                  object
consumo_prom_mixto      float64
consumo_total           float64
consumo_prom            float64
consumo_prom_no_dom     float64
bimestre                  int64
consumo_total_no_dom    float64
indice_des               object
latitud                 float64
longitud                float64
dtype: object

transformers = [('impute_consumo_total_mixto', SimpleImputer(strategy="median"), ['consumo_total_mixto']),
                ('impute_consumo_prom_dom', SimpleImputer(strategy="median"), ['consumo_prom_dom']),
                ('impute_consumo_total_dom', SimpleImputer(strategy="median"), ['consumo_total_dom']),
                ('impute_consumo_prom_mixto', SimpleImputer(strategy="median"), ['consumo_prom_mixto'])       
               ]

col_trans = ColumnTransformer(transformers, remainder="drop", n_jobs=-1)
col_trans.fit(datos_eda)

In [ ]:
transformers = [('impute_consumo_total_mixto', SimpleImputer(strategy="median"), ['consumo_total_mixto']),
                ('impute_consumo_prom_dom', SimpleImputer(strategy="median"), ['consumo_prom_dom']),
               ('impute_consumo_total_dom', SimpleImputer(strategy="median"), ['consumo_total_dom']),
               ('impute_consumo_prom_mixto', SimpleImputer(strategy="median"), ['consumo_prom_mixto'])]

col_trans = ColumnTransformer(transformers, remainder="passthrough", n_jobs=-1, verbose=True)
col_trans.fit(df3)

u = col_trans.transform(df3)
u[0:5,]
#print(u[0:5,])
df4 = pd.DataFrame(u, 
             columns=['consumo_total_mixto','consumo_prom_dom', 'consumo_total_dom','consumo_prom_mixto','latitud','longitud','anio',  
       'alcaldia','consumo_total', 'consumo_prom','consumo_prom_no_dom', 'bimestre', 'consumo_total_no_dom', 'indice_des'])
df4.head()

In [21]:
X = datos_eda.consumo_total_mixto.values.reshape(datos_eda.shape[0],1)